File: battle_engine/app/contracts/consumers/select_leads_contract.rb
# frozen_string_literal: true

module Contracts
  module Consumers
    class SelectLeadsContract < Dry::Validation::Contract
      params do
        required(:battle_id).filled(:string)
        required(:players).array(:hash) do
          required(:player_id).filled(:string)
          required(:lead).filled(:integer)
        end
      end
    end
  end
end

File: battle_real_time/lib/battle_real_time/contracts/publishers/battle_actions/select_leads_contract.ex
defmodule BattleRealTime.Contracts.Publishers.BattleActions.SelectLeadsContract do
  use BattleRealTime.Contracts.Contract

  @primary_key false
  embedded_schema do
    field(:battle_id, :string)

    embeds_many :players, PlayerLead, primary_key: false do
      field(:player_id, :string)
      field(:lead, :integer)
    end
  end

  def changeset(struct, params) do
    struct
    |> cast(params, [:battle_id])
    |> cast_embed(:players, required: true, with: &player_lead_changeset/2)
    |> validate_required([:battle_id])
  end

  defp player_lead_changeset(struct, params) do
    struct
    |> cast(params, [:player_id, :lead])
    |> validate_required([:player_id, :lead])
  end
end
